### PR TITLE
Tutorial updates

### PR DESCRIPTION
**Pull request by Kilo59**
Updated the tutorial docs & examples to be slightly more modern & expressive.

There's a couple of cases where the tutorial code doesn't work as presented (even with imports https://github.com/dagster-io/dagster/issues/3975) and one would have to open up the example doc snippets to get it to work.
I started to update these to work "as is", but thought it's probably better to do that in a separate PR.
~~So I may drop the commits that have those changes.~~

## Summary

There are basically 3 different types of updates I made.

1. Replaced all uses of `os` with `pathlib`.
    * More modern, [cross-platform friendlier](https://treyhunner.com/2019/01/no-really-pathlib-is-great/#Normalizing_file_paths_shouldn%E2%80%99t_be_your_concern), more expressive way to work with paths.
    *  https://treyhunner.com/2018/12/why-you-should-be-using-pathlib/
    * https://treyhunner.com/2019/01/no-really-pathlib-is-great/
    * https://realpython.com/python-pathlib/#the-problem-with-python-file-path-handling
    * https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module

2. Updated .format with f-string (where appropriate)
   * The existing documentation switches between f-strings and `.format()` 
   * In some cases a switch to `f-string` was done but someone left an extraneous `.format()` call with no values.
   * In other cases it looks like the `.format()` method actually helps the code self-document itself better so I left it alone. `.format(some_descriptive_variable_name=unclear + something_else_unclear)`

3. Use `enumerate()` instead of `for i in range(len(value))`
   * self-explanatory 

## Checklist

- [x] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: docs/content/tutorial/advanced-tutorial/types.mdx [hunk 1]
```diff
@@ -22,8 +22,8 @@ Let's look back at our simple `read_csv` solid.
 
 ```python file=/intro_tutorial/basics/e04_quality/inputs_typed.py startafter=start_inputs_typed_marker_0 endbefore=end_inputs_typed_marker_0
 @solid
-def read_csv(context, csv_path: str):
-    csv_path = os.path.join(os.path.dirname(__file__), csv_path)
```
Comment (Kilo59) [inline docs/content/tutorial/advanced-tutorial/types.mdx:26]: Here's a case where the tutorial section doesn't seem to work "as is" without adding in some of the elided content from the docs snippets. 
I'm going to revert this from this PR (not counting the pathlib use).

File: examples/docs_snippets/docs_snippets/intro_tutorial/advanced/materializations/materializations.py [hunk 2]
```diff
@@ -38,10 +38,10 @@ def sort_by_calories(context, cereals):
...
     )
     fieldnames = list(sorted_cereals[0].keys())
-    sorted_cereals_csv_path = os.path.abspath(
-        "output/calories_sorted_{run_id}.csv".format(run_id=context.run_id)
-    )
-    os.makedirs(os.path.dirname(sorted_cereals_csv_path), exist_ok=True)
+    sorted_cereals_csv_path = pathlib.Path(
+        f"output/calories_sorted_{context.run_id}.csv"
+    ).resolve()
+    sorted_cereals_csv_path.parent.mkdir(exist_ok=True)
```
Comment (Kilo59) [inline examples/docs_snippets/docs_snippets/intro_tutorial/advanced/materializations/materializations.py:44]: https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir

File: examples/docs_snippets/docs_snippets/intro_tutorial/advanced/materializations/output_materialization.py [hunk 3]
```diff
@@ -47,10 +47,8 @@ def less_simple_data_frame_loader(context, config):
 )
 def less_simple_data_frame_materializer(context, config, value):
     # Materialize LessSimpleDataFrame into a csv file
-    csv_path = os.path.join(
-        os.path.dirname(__file__), os.path.abspath(config["csv"]["path"])
-    )
-    os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+    csv_path = pathlib.Path(__file__).parent.joinpath(config["csv"]["path"]).resolve()
```
Comment (Kilo59) [inline examples/docs_snippets/docs_snippets/intro_tutorial/advanced/materializations/output_materialization.py:50]: Or maybe...
```python
csv_path = pathlib.Path(__file__).parent / pathlib.Path((config["csv"]["path"]).resolve()
```

File: docs/content/tutorial/advanced-tutorial/types.mdx [hunk 4]
```diff
@@ -152,31 +152,29 @@ def less_simple_data_frame_type_check(_, value):
...
-                "{type_}"
-            ).format(type_=type(value)),
+                f"LessSimpleDataFrame should be a list of dicts, got {type(value)}"
+            ),
         )
 
     fields = [field for field in value[0].keys()]
 
-    for i in range(len(value)):
-        row = value[i]
```
Comment (Kilo59) [inline docs/content/tutorial/advanced-tutorial/types.mdx:163]: `enumerate()` is designed for exactly this use case.

**Comment by alangenfeld** [comment]
hey thanks for the PR and sorry for the delay. Most of these are artifacts from supporting py27 which we dropped not that long ago. 

I think most of the unrelated errors would go away with a rebase
`make black` in the root should resolve the black failure

heres some info on the docs snapshot 
https://github.com/dagster-io/dagster/tree/master/docs#code-snippets-literal-includes

**Comment by Kilo59** [comment]
Not going to do this over.

**Pull request closed without merging**